### ISSUE TITLE
Migrating from v1.0.2 to v2.0.0 bugs the shadows

### ISSUE THREAD

**Issue by ultWorld** [StrandedKitty/three-csm#20]
![image](https://user-images.githubusercontent.com/104983217/193844327-601b5e62-88e6-4295-986f-6b83f61f4850.png)

Hi, i am working on a fork of [sketchbook](https://github.com/swift502/Sketchbook) and want to migrate to latest versions of the libraries.

Do you have any idea on what can i do to fix?

**Comment by ultWorld** [comment]
@swift502 i tried to test a lot on customSplitCallback but unfortunatly i feel that is not the problem anymore. :/

P.S. : This is Vivasvan from your Discord. :)

**Comment by StrandedKitty** [comment]
Hi, could you please provide a more detailed explanation of your problem? What exactly is broken, what version of three.js you are using in your project, what have you already tried to fix it, etc. Also it would be nice if you could provide a demo.

**Comment by vivasvan1** [comment]
There are 2 issues according to me.

1) the character and object start casting very faded and weak shadows on the floor after the migration.

2) there seems to be leak from the edges of the walls. ( Not sure if this is normal or not )

Both are visible in the image.

I am using v0.145.0 for three js https://www.npmjs.com/package/three

Thing i tried:

1) I tried to removing all other light sources keeping only the directional light. (The above image)

2) i tried multiple different customSplitCallback which didn't work.

3) i played around with all other params of CSM which doesn't seem to fix both the issues. ( Reducing the light intensity minimises the issue 2. But its still there. )


Will send you a demo link shortly.

**Comment by swift502** [comment]
I believe the only issue is just an extremely large shadow texture size.

- that magnifies the edge leaks, which are always present, but are usually small enough to not be as visible
- and it causes objects to cast "faded" and "weak" shadows, because they have very few shadow pixels to work with

How far is the clipping distance of the CSM? Can it maybe be reduced and the shadow map resolution increased?

**Comment by swift502** [comment]
Another idea: have you been changing the glb models? are you sure you aren't importing another light sources from the glbs?
It kinda looks like there's multiple overlapping shadows, one sharp, and other very blurry:

![image](https://user-images.githubusercontent.com/24359130/193865496-ccb7172a-2e85-4f5e-a4f5-286cc669f9cd.png)

In the vanilla sketchbook demo, nothing is ever perfectly black, like the tunnel shadow in your picture. I don't see how that could happen, except if multiple shadow casters were combined to turn the tunnel floor completely black.

**Comment by StrandedKitty** [comment]
> 2) there seems to be leak from the edges of the walls. ( Not sure if this is normal or not )

This is most likely related to `shadowBias`, try using a larger value. Shadow bias must be adjusted manually for your scene, so try playing with it and see if it helps to reduce leaks.

**Comment by ultWorld** [comment]
Thank you guys for your reply.

> Another idea: have you been changing the glb models? are you sure you aren't importing another light sources from the glbs? It kinda looks like there's multiple overlapping shadows, one sharp, and other very blurry:

@swift502  Yeah I am sure there is not light i removed the hemi-light and i am using your original glb. 

> How far is the clipping distance of the CSM?

I tried everything from 0 to camera.far( i.e. 1010) ....weirdly this parameter doesnt seem to change anything.

> This is most likely related to shadowBias, try using a larger value. Shadow bias must be adjusted manually for your scene, so try playing with it and see if it helps to reduce leaks.

shadowBias: 
0.001 - seems to completely removes all shadows above and below tunnel.
0.0001 - Makes the tunnel shadow show on top and also no player shadow on floor.  
![image](https://user-images.githubusercontent.com/104983217/194106963-55a021ca-7efe-4e41-8e27-30838ee653e0.png)
0.00001 - reduces the effect but still like 0.0001
0.000001 - futher reduces the effect of tunnel shadow on top but still no character shadow.

![image](https://user-images.githubusercontent.com/104983217/194105206-597a67da-6434-4701-bed9-e817b096e4a2.png)

**Comment by ultWorld** [comment]
I wrote this small demo .... why is the shadow still blur in this?

https://codepen.io/vivasvan100/pen/BaxPeLb?editors=1010

not sure but it could be the same effect.

**Comment by swift502** [comment]
@ultWorld  I think it works just fine. Using the settings from my Sketchbook demo, you get shadows as you see them in my demo.

![image](https://user-images.githubusercontent.com/24359130/194234073-3f097ebe-cc5b-428a-a9fc-3c363024dbdf.png)

```js
// I had to slightly modify this function, the "target" parameter is new in v2
let splitsCallback = (amount, near, far, target) =>
{
	for (let i = amount - 1; i >= 0; i--)
	{
		target.push(Math.pow(1 / 4, i));
	}
};

let csm = new THREE.CSM({
	maxFar: 250,
	cascades: 3,
	shadowMapSize: 2048,
	lightDirection: new THREE.Vector3(1, -1, 1).normalize(),
	camera: camera,
	parent: scene,
	mode: 'custom',
	customSplitsCallback: splitsCallback,
	lightFar: 5000,
	lightNear: 1,
	shadowBias: 0,
});
csm.fade = true;
```

As I suspected, your maxFar is far too far (funny, I know), set maxFar to 10 and see how sharp the shadows immediately become.

![image](https://user-images.githubusercontent.com/24359130/194234716-5be9e6a2-91bc-469f-b66a-175bd70dadc9.png)

shadowBias really isn't the main issue here, of course you'll have edge leaks if a shadow pixel is as big as the entire wall it's cast from. Same with the character model.

So obviously CSM works, now it's about how you allocate the splits, and how much resolution you give the shadow maps.
I believe practical splits don't offer the right split distribution, which is why I always used my own splits.
You can also increase the shadowMapSize, with an understandable performance penalty.

![image](https://user-images.githubusercontent.com/24359130/194236350-16e62b8b-ffef-4904-8678-a1319be64288.png)

Finally, in my Sketchbook demo I got away with the highest shadow resolution I could, while already sacrificing a lot of performance. @StrandedKitty always emphasized how "unoptimized" this system essentially is, and that it's performance penalty might always be quite high compared to the results.

**Comment by swift502** [comment]
@vivasvan1 Now, are these demo settings the same as in your Sketchbook fork?
What is your maxFar in the Sketchbook fork?

**Comment by swift502** [comment]
PS: While you can tie shadow maxFar to camera.far, I think most games don't do that. Usually the shadows can disappear quite a bit before camera.far, and it still looks acceptable. And you save lots of shadow resolution and performance.

![image](https://user-images.githubusercontent.com/24359130/194247932-062a4772-f900-4907-8fc2-8ae813fdb34c.png)

Maybe modern games do it differently, but traditionally I think it's just done like this.
Just because in the code pen you set `maxFar: camera.far`, but maybe that's borrowed from the official CSM examples.

**Comment by swift502** [comment]
PPS: The reason the official [Basic](https://strandedkitty.me/three-csm/examples/basic/) example works with a 1000 maxFar is because it must be absolutely massive compared to your cube in the codepen, or even the Sketchbook world.

So scaling up your 3D scene should work just as well as lowering the distance parameters in the CSM.

That's my theory anyways.

**Comment by ultWorld** [comment]
@swift502  Thank you so much. It did work. Amazing work.

> As I suspected, your maxFar is far too far (funny, I know), set maxFar to 10 and see how sharp the shadows immediately become.

hahah. Yes, this is correct. This is what was going on in both the demo and my fork. reducing the maxFar to 300 from camera.far worked. Also the leak on the edge seems to be reduced quite a bit. 

![image](https://user-images.githubusercontent.com/104983217/194266635-e7cd88fb-4c83-445a-b9b1-02dfec312f6f.png)

![image](https://user-images.githubusercontent.com/104983217/194270546-4b614743-8c78-49e5-a8ec-d822fd8ad518.png)


> You can also increase the shadowMapSize, with an understandable performance penalty.

This is also helps. I changed it 2048 -> 4096. and the shadows look pretty much how i desired. On my laptop the performace drop is hardly noticible but maybe will set it 2048 or even less for mobile devices.

### This works now but would it be possible dark shadows like the once you have in your original work? 

![image](https://user-images.githubusercontent.com/104983217/194274681-af83b46d-d1be-41c4-8c7a-a45949e91b92.png)


And again thank you so much to both of you. Really appriciate your help <3 :+1: 

P.S. 

> @vivasvan1 Now, are these demo settings the same as in your Sketchbook fork?
What is your maxFar in the Sketchbook fork?

This was the perfect question. Its crazy to me how you figured it out only from this :sunglasses:

**Comment by swift502** [comment]
Cool!

> This works now but would it be possible dark shadows like the once you have in your original work?

What's odd to me that in some places the shadow darkness is correct right?

![image](https://user-images.githubusercontent.com/24359130/194297501-b1c6fb6e-db97-4db1-a739-c09dba38ba6c.png)

Generally for darker shadows, you would just decrease the hemi light intensity.
But you seem to have a "smooth edge" to your shadows, so they go from light to dark at the edges. And I don't know what that's about... Maybe a new feature of the CSM system? 😄

**Comment by swift502** [comment]
Oh, well, could it be the new fading feature? Do the shadows look different when you set `fade = false`?
It could just be interpolating between the two closest cascades, I imagine that would look something like this. Just an idea.

**Comment by StrandedKitty** [comment]
@ultWorld Hi, did you manage to fix these problems with shadows?

**Comment by vivasvan1** [comment]
@StrandedKitty yes i did manage to fix it ... check the ongoing PR on the Sketchbook's repository.
